### PR TITLE
[indexer] Fix standalone relation reads

### PR DESCRIPTION
**Pull request by biodranik**
`FeaturesVectorTest` previously constructed `FeaturesVector` before loading its relation-offset table. `SharedLoadInfo` therefore did not open the relations reader, even though the table pointer was assigned afterward. Calling `GetRelation()` then asserted in Debug and dereferenced an empty optional in Release.

Make `FeaturesVectorTest` own its dependencies and pass them when constructing `FeaturesVector`. Keep the relation table immutable after construction, fail explicitly if its reader is unavailable, and add a regression test for standalone relation reading.

Found while investigating #13434 and #13435; this PR does not fix those rendering issues.

Tested: `indexer_tests --filter="FeaturesVectorTest_.*"`

### CODE REVIEW

File: libs/indexer/indexer_tests/features_vector_test.cpp [hunk 1]
```diff
@@ -1,20 +1,72 @@
...
+  }
+
+  vector<uint8_t> relations;
+  MemWriter writer(relations);
+  TestRouteRelation().Write(writer);
+  if (datHeader.m_version >= feature::DatSectionHeader::Version::V2)
+  {
+    WriteVarUint(writer, uint32_t{0});  // Feature members.
+    WriteToSink(writer, uint8_t{0});    // Relation members and parents flags.
+  }
```
Comment (biodranik) [inline libs/indexer/indexer_tests/features_vector_test.cpp]: This branch is dead with the current test data: `data/minsk-pass.mwm` has dat section version **V0** (its `features` section starts at file offset 64 and the first byte there is `0`), so `m_version >= V2` is never true.

Consequence: `SharedLoadInfo::GetRelation()` takes the `< V2` path (`RouteRelationBase::Read` only), and `TEST(relation.GetMembers().empty(), ())` below is vacuous — the V2 member decoding (`ReadVarInt` delta loop in `RouteRelation::Read`) is never exercised by this test.

Either drop the conditional and the members check (the test then honestly covers only what actually regressed — base relation reads through `FeaturesVectorTest`), or regenerate the test mwm as V2 and write one real member so the delta decoding is covered.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: generator/dumper.cpp [hunk 2]
```diff
@@ -108,6 +111,72 @@ void DumpTypes(string const & fPath)
...
+    cout << "No relations sections in " << fPath << endl;
+    return;
+  }
+
+  auto const relTable = feature::FeaturesOffsetsTable::Load(cont, RELATION_OFFSETS_FILE_TAG);
+  std::unique_ptr<feature::FeaturesOffsetsTable> osmIds;
+  if (cont.IsExist(RELATION_OSMIDS_FILE_TAG))
+    osmIds = feature::FeaturesOffsetsTable::Load(cont, RELATION_OSMIDS_FILE_TAG);
+
+  size_t const count = relTable->size();
```
Comment (biodranik) [inline generator/dumper.cpp]: `FeaturesVectorTest` already loaded this exact section into `m_relTable` (that's the point of the first commit) — this is a second elias-fano decode of the same data just to get `size()`.

Add an accessor next to the existing `GetNumFeatures()`:

```cpp
// features_vector.hpp
size_t GetNumRelations() const;
// features_vector.cpp
size_t FeaturesVector::GetNumRelations() const
{
  return m_loadInfo.m_relTable ? m_loadInfo.m_relTable->size() : 0;
}
```

then the `IsExist()` check and the `Load()` call here collapse to:

```cpp
size_t const count = fv.GetVector().GetNumRelations();
if (count == 0)
{
  cout << "No relations in " << fPath << endl;
  return;
}
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: generator/dumper.cpp [hunk 3]
```diff
@@ -108,6 +111,72 @@ void DumpTypes(string const & fPath)
...
+
+  for (uint32_t i = 0; i < count; ++i)
+  {
+    auto const rel = fv.GetVector().GetRelation(i);
+    string const name(rel.GetDefaultName());
+
+    if (rel.GetType() != feature::RouteRelationBase::Type::Subway)
+      continue;
+    if (filter != "all" && name.find(filter) == string::npos && rel.GetRef() != filter)
+      continue;
```
Comment (biodranik) [inline generator/dumper.cpp]: Two full passes over all relations. The histogram pass deserializes every relation completely (name multilang blob, network/ref, all params, member vector) just to read one type byte. A single pass is half the I/O and fewer lines: accumulate `typeCounts` inside the detail loop and print the histogram at the end.

Two more in the same block:

* `string const name(rel.GetDefaultName())` allocates before the type filter rejects the relation. `std::string_view` works for both `find()` and `operator<<`, and the copy is unnecessary anyway since `rel` outlives it.
* `Type::Subway` is hardcoded, so the flag cannot investigate the other 14 route types (#13434/#13435 are subway-specific, but a debug tool that only sees subways will need editing next time). Matching `filter` against `ToString(rel.GetType())` as well removes this line and makes `--dump_relations=Subway` do exactly what the hardcode does today.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/indexer/features_vector.cpp [hunk 4]
```diff
@@ -50,20 +50,12 @@ FeaturesVectorTest::FeaturesVectorTest(std::string const & filePath)
...
-
-  if (m_cont.IsExist(METADATA_FILE_TAG))
-    m_vector.m_loadInfo.m_metaDeserializer = indexer::MetadataDeserializer::Load(m_cont).release();
-}
+  , m_ftTable(feature::FeaturesOffsetsTable::Load(m_cont, FEATURE_OFFSETS_FILE_TAG))
+  , m_relTable(m_cont.IsExist(RELATION_OFFSETS_FILE_TAG)
+                   ? feature::FeaturesOffsetsTable::Load(m_cont, RELATION_OFFSETS_FILE_TAG)
+                   : nullptr)
+  , m_metaDeserializer(m_cont.IsExist(METADATA_FILE_TAG) ? indexer::MetadataDeserializer::Load(m_cont) : nullptr)
+  , m_vector(m_cont, m_header, m_ftTable.get(), m_relTable.get(), m_metaDeserializer.get())
```
Comment (biodranik) [inline libs/indexer/features_vector.cpp:58]: Good fix — the root cause is that `SharedLoadInfo` opens `m_relsReader` **only if** `relTable != nullptr` at construction time, so the old post-hoc `m_loadInfo.m_relTable = ...` left `m_relsReader` empty and `GetRelation()` dereferenced an empty `optional` in Release.

Dropping `friend class FeaturesVectorTest` closes this hole, but `SharedLoadInfo::m_relTable` / `m_metaDeserializer` are still public and mutable, so the same latent UB can be reintroduced. After this PR neither is ever assigned post-construction (only `m_version` is, from `InitRecordsReader`), so they can be locked down for free:

```cpp
public:
  feature::FeaturesOffsetsTable const * const m_relTable;
  indexer::MetadataDeserializer * const m_metaDeserializer;  // pointee stays mutable
  feature::DatSectionHeader::Version m_version;
```

`SharedLoadInfo` is already `DISALLOW_COPY_AND_MOVE`, so `const` members cost nothing and the compiler now enforces the "reader matches table" invariant.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by biodranik — COMMENTED** [review]
The root cause is confirmed: `SharedLoadInfo`'s ctor opens `m_relsReader` only when `relTable != nullptr`, and the old `FeaturesVectorTest` assigned `m_loadInfo.m_relTable` *after* that ctor ran — so `m_relTable` was non-null while `m_relsReader` stayed `nullopt`, and every `GetRelation()`/`ReadRelation()` hit `ASSERT` in Debug and dereferenced an empty `optional` in Release. `MwmValue`-based paths (`feature_source.cpp`, `mwm_context.cpp`) pass the tables through the ctor, so `FeaturesVectorTest` was indeed the only broken construction site.

Findings beyond the inline ones:

**1. Make the footgun impossible — `SharedLoadInfo::m_relTable` should be private.**
It is a public non-const member, which is exactly what allowed the post-construction assignment that caused this bug. Grep shows it is referenced *only* inside `shared_load_info.hpp/.cpp`; moving it into the private section next to `m_relsReader` turns this whole bug class into a compile error, at zero cost. (`m_version` and `m_metaDeserializer` must stay public — they are written/read from `features_vector.cpp` and `feature.cpp`.)

**2. `ASSERT(m_relsReader, ())` → `CHECK` in `shared_load_info.cpp:41,47`.**
Violating it is silent UB in Release (`*empty_optional`), which is what the branch name refers to. Per the project's fail-fast policy this is a condition that must hold in Release too; one branch per relation read is not measurable next to the elias-fano lookup and the deserialization that follows.

**3. `DebugPrint(DatSectionHeader::Version)` is wrong for V1/V2** (`libs/indexer/dat_section_header.hpp`) — the `default:` arm returns `"Latest(V1)"`, so a V2 header prints as V1. The new test now depends on version semantics and `DatSectionHeader::Read()` reports through this function on `CHECK` failure, so it is worth fixing here:
```cpp
case DatSectionHeader::Version::V0: return "V0";
case DatSectionHeader::Version::V1: return "V1";
case DatSectionHeader::Version::V2: return "V2";
```

**4. The dumper's leading comment.** `// Temporary debug dump for route relations (issues #13434, #13435 investigation).` describes the reason it was written rather than what the code does; once those issues close it is stale. If the tool stays, say what it prints; if it goes, the comment goes with it.

On the design question in the description: keeping the tables owned by `FeaturesVectorTest` and passing raw pointers into `FeaturesVector` is the right call — `MwmValue` shares `m_ftTable`/`m_relTable` between values via `weak_ptr`, so `FeaturesVector` must stay non-owning. An owning `FeaturesVector` ctor overload would force that sharing to be duplicated. Removing `friend class FeaturesVectorTest` is a clean net win.

Regarding `--dump_relations` itself: it is genuinely useful for the relation work (nothing else can show member geometry types per relation), and it costs 60 lines in a file that already exists for exactly this. My suggestion is to keep it but generalize past subways (see the inline comment) — a debug tool that needs a recompile for the next route type is not worth keeping.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: generator/dumper.cpp [hunk 5]
```diff
@@ -108,6 +111,72 @@ void DumpTypes(string const & fPath)
...
+    return;
+  }
+
+  auto const relTable = feature::FeaturesOffsetsTable::Load(cont, RELATION_OFFSETS_FILE_TAG);
+  std::unique_ptr<feature::FeaturesOffsetsTable> osmIds;
+  if (cont.IsExist(RELATION_OSMIDS_FILE_TAG))
+    osmIds = feature::FeaturesOffsetsTable::Load(cont, RELATION_OSMIDS_FILE_TAG);
+
+  size_t const count = relTable->size();
+  cout << "Relations count: " << count << "; rel_osmids: " << (osmIds ? "present" : "MISSING") << endl;
```
Comment (biodranik) [inline generator/dumper.cpp]: This `MISSING` branch will fire on **every** mwm, and that looks like the direct cause of #13435.

`grep -r rel_osmids` over the whole tree returns only 3 places: the `#define` in `defines.hpp`, the two readers in `libs/map/relation_track.cpp` (:294, :333), and this dump. Nothing in `generator/` (or `tools/python/maps_generator/`) ever writes the section. So in `RelationTrackBuilder::AppendNeighbourMembers()`:

```cpp
if (cont.IsExist(RELATION_OSMIDS_FILE_TAG)) { ... }
else
  return;                                  // <-- always taken
```

the BFS over neighbouring mwms never starts, and `TryAppendFromMwm()` is dead code — a metro line crossing an mwm border can never be joined, exactly as reported for Paris M11.

Worth confirming with this tool on a real map before spending more time on the reader side; if it prints `MISSING` everywhere, the fix belongs in the generator (emit `rel_osmids` alongside `rel_offs`), not in `relation_track.cpp`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: generator/dumper.cpp [hunk 6]
```diff
@@ -108,6 +111,72 @@ void DumpTypes(string const & fPath)
...
+    cout << "  type " << t << ": " << c << endl;
+
+  for (uint32_t i = 0; i < count; ++i)
+  {
+    auto const rel = fv.GetVector().GetRelation(i);
+    string const name(rel.GetDefaultName());
+
+    if (rel.GetType() != feature::RouteRelationBase::Type::Subway)
+      continue;
+    if (filter != "all" && name.find(filter) == string::npos && rel.GetRef() != filter)
```
Comment (biodranik) [inline generator/dumper.cpp]: The two halves of the filter behave differently: substring match on the name, exact match on the ref. So `--dump_relations=11` finds every relation with `11` anywhere in the name but misses `ref='M11'`. Make both substrings:

```cpp
if (filter != "all" && name.find(filter) == string::npos && rel.GetRef().find(filter) == string::npos)
  continue;
```

Also note `GetDefaultName()` is the default-locale name only, so filtering a map by its local metro line name will silently return nothing — worth one word in the flag help (`generator_tool.cpp:162`), which currently says "subway route relations" for a flag named `dump_relations`.

Nit, line 166: `default: break;` in the `GeomType` switch only covers `Undefined`, which for a stored feature is a bug. Spelling `case feature::GeomType::Undefined: break;` keeps `-Wswitch` useful and matches the fail-fast style.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by biodranik — COMMENTED** [review]
Follow-up pass (the earlier review already covers the dead V2 branch in the test, the duplicate `rel_offs` load, the double relation pass and the hardcoded `Subway`).

Two things not raised yet — the `rel_osmids` one is the interesting find, since it appears to answer #13435 outright.

Nothing to flag on the `FeaturesVectorTest` fix itself beyond the inline note: member init order matches declaration order, the out-of-line `= default` dtor is required (both `FeaturesOffsetsTable` and `MetadataDeserializer` are incomplete in the header), and all 13 `FeaturesVectorTest` call sites only read through `GetVector()`/`GetContainer()`, so removing the `friend` breaks nothing. Verified that `MwmValue::SetTable()` populates the tables before any `FeaturesVector` is built, so production paths were never affected by the original bug.

No translations or search terms in this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by vng** [comment]
- I'm not clear on the first commit fix. Everything was encapsulated without additional (duplicated) unique_ptr fields. "Outlive"? The before and after m_relTable lifetime <= FeaturesVectorTest lifetime.
- The second commit looks strange from the first lines with duplicating RELATION_OFFSETS_FILE_TAG reading.

**Comment by biodranik** [comment]
Thanks, you are right that “outlive” was misleading; this was not primarily a use-after-free.

The old `FeaturesVectorTest` constructed `FeaturesVector` with `relTable == nullptr`. During that construction, `SharedLoadInfo` only initialized `m_relsReader` when the relation table was already non-null. The constructor body later assigned `m_relTable`, but `m_relsReader` remained an empty `std::optional`.

Consequently, `GetRelation()` hit `ASSERT(m_relsReader)` in Debug. In Release the assertion is compiled out, and `*m_relsReader` dereferenced the empty optional, which is UB. This path was latent in `FeaturesVectorTest`; normal `MwmValue` paths already pass the table through the constructor.

The new `unique_ptr` fields are not duplicate tables: they explicitly own the objects that `FeaturesVector` references non-owningly and make them available during its construction.

Agreed about the duplicated `RELATION_OFFSETS_FILE_TAG` read. I dropped the entire temporary dumper commit and kept this PR focused on the standalone reader fix. I also made `SharedLoadInfo::m_relTable` private and immutable and changed the relation-reader guards to `CHECK` so invalid Release use fails explicitly instead of reaching UB.

File: libs/indexer/shared_load_info.cpp [hunk 7]
```diff
@@ -37,13 +37,13 @@ SharedLoadInfo::Reader SharedLoadInfo::GetTrianglesReader(size_t ind) const
 
 RelationReader SharedLoadInfo::ReadRelation(uint32_t id) const
 {
-  ASSERT(m_relsReader, ());
+  CHECK(m_relsReader, ());
```
Comment (biodranik) [inline libs/indexer/shared_load_info.cpp]: @vng leave ASSERT here and below?

Comment (vng) [reply libs/indexer/shared_load_info.cpp]: Yes, the test is enough.

**Review by vng — APPROVED** [review]
